### PR TITLE
Add `ArrayFlat` type

### LINKED ISSUE #1075: Proposal: `Flat` (equivalent to `Array.prototype.flat()`)
### Type description + examples

I'd like to propose something like `Array.prototype.flat()` but for types.

### Type source

```ts
type Flat<T extends any[][]> = T extends (infer U)[][] ? U : never;

type Input = [
  ['A', 'B'],
  ['C'],
]

type Output = Flat<Input>
// => ['A', 'B', 'C']
```

### Search existing types and issues first

- [x] I tried my best to look for it

### PR DESCRIPTION
**Pull request by Emiyaaaaa**
- closed #1075

### CODE REVIEW

File: test-d/array-flat.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+expectType<ArrayFlat<readonly [readonly [1, 2], readonly [3, 4]]>>([1, 2, 3, 4]);
+
+// Edge cases
+expectType<ArrayFlat<Array<[]>>>([]);
+expectType<ArrayFlat<[[]]>>([]);
+expectType<ArrayFlat<[[], []]>>([]);
+expectType<ArrayFlat<undefined[]>>(null! as undefined[]);
+expectType<ArrayFlat<any[]>>(null! as any[]);
+expectType<ArrayFlat<unknown[]>>(null! as unknown[]);
+expectType<ArrayFlat<never[]>>(null! as never[]);
```
Comment (sindresorhus) [inline test-d/array-flat.ts:60]: Some more potential tests (AI generated, but I looked over them):

```js
// Test specifically for non-tuple array handling (Array<T> vs [T, T])
type GenericNumberArrays = Array<Array<number>>;
const genericArrayTest: ArrayFlat<GenericNumberArrays> = [1, 2, 3, 4, 5];

// Test for deeply nested non-tuple arrays with specific depths
type DeepGenericArray = Array<Array<Array<Array<number>>>>;
const deepGenericTest1: ArrayFlat<DeepGenericArray, 2> = [[1, 2], [3, 4]];
const deepGenericTest2: ArrayFlat<DeepGenericArray, 3> = [1, 2, 3, 4];

// Test for mixed generic arrays and tuples
type MixedGenericAndTuple = Array<[number, string]>;
const mixedGenericTupleTest: ArrayFlat<MixedGenericAndTuple> = [1, 'a', 2, 'b'];

// Test for array with optional elements in nested structure
type NestedOptional = [number, [string?, number?][]];
const nestedOptionalTest: ArrayFlat<NestedOptional> = [1, undefined, 2, 'a', 3];

// Test for arrays with rest elements in nested structure
type NestedRest = [string, [...number[]][]];
const nestedRestTest: ArrayFlat<NestedRest> = ['a', 1, 2, 3, 4, 5];

// Test for flattening arrays with union types in nested structures
type NestedUnion = Array<Array<string | number>>;
const nestedUnionTest: ArrayFlat<NestedUnion> = ['a', 1, 'b', 2];

// Test for empty array in complex structure
type ComplexWithEmpty = [number, Array<[]>, string];
const complexEmptyTest: ArrayFlat<ComplexWithEmpty> = [1, 'string'];

// Test for array with undefined/null elements
type ArrayWithNullish = [number, [undefined, null]];
const nullishTest: ArrayFlat<ArrayWithNullish> = [1, undefined, null];

// Test for array with mixed depth elements
type MixedDepthArray = [number, Array<string>, [[boolean]]];
const mixedDepthTest: ArrayFlat<MixedDepthArray, 1> = [1, 'a', 'b', [true]];

// Test for readonly nested arrays with different depths
type ReadonlyNestedComplex = readonly [number, readonly (readonly string[])[]];
const readonlyComplexTest: ArrayFlat<ReadonlyNestedComplex, 2> = [1, 'a', 'b', 'c'];

// Test for recursive flattening with non-array elements
type RecursiveWithNonArray = [number, [string, {a: number}]];
const recursiveNonArrayTest: ArrayFlat<RecursiveWithNonArray> = [1, 'string', {a: 42}];
```

File: test-d/array-flat.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+// Test with different element types
+expectType<ArrayFlat<[string, [number, boolean]]>>([null! as string, null! as number, null! as boolean]);
+expectType<ArrayFlat<[{a: number}, [{b: string}]]>>([null! as {a: number}, null! as {b: string}]);
+
+// Test with union types
+expectType<ArrayFlat<[1, 2] | [[3, 4]]>>([1, 2] as [1, 2] | [3, 4]);
+expectType<ArrayFlat<[1, [2]] | [[3], 4]>>([1, 2] as [1, 2] | [3, 4]);
+
+expectType<ArrayFlat<[...Array<[number, string]>]>>(null! as [number?, string?, number?, string?, number?, string?, number?, string?, number?, string?, number?, string?, number?, string?, number?, string?, number?, string?, number?, string?]);
```
Comment (som-sm) [inline test-d/array-flat.ts:37]: But the output of `[number?, string?, number?, ...]` wouldn't work if the input were `[...Array<[number, string?]`.

Consider this example:
```ts
const input: [...Array<[number, string?]>] = [[1], [3], [5]];
const output: ArrayFlat<[...Array<[number, string?]>]> = [1, 3, 5]; // Errors, but it shouldn't
```

File: test-d/array-flat.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+expectType<ArrayFlat<[1, [2, ...Array<3>], 4]>>([1, 2, ...(null! as Array<3>), 4]);
+
+// Test with mixed arrays and tuples
+expectType<ArrayFlat<[1, number[], 3]>>([1, ...(null! as number[]), 3]);
```
Comment (som-sm) [inline test-d/array-flat.ts:41]: Add these cases:

```ts
ArrayFlat<[1, number[], string[], true]>
ArrayFlat<[1, number[], true, string[]]>
ArrayFlat<[string[], [...boolean[], string]]>
```

Comment (som-sm) [inline test-d/array-flat.ts:41]: Fails in this case: `ArrayFlat<[...number[], string[]]>`

Returns `[...number[], string[]]` which isn't correct, should probably return `(number | string)[]` instead.

```ts
const input: [...number[], string[]] = [1, 2, 3, ['a', 'b', 'c']];
const output: ArrayFlat<[...number[], string[]]> = [1, 2, 3, 'a', 'b', 'c']; // Errors, but it shouldn't
```

File: test-d/array-flat.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+// Test for deeply nested non-tuple arrays with specific depths
+type DeepGenericArray = number[][][][];
+expectAssignable<ArrayFlat<DeepGenericArray, 2>>([[1, 2], [3, 4]]);
+expectAssignable<ArrayFlat<DeepGenericArray, 3>>([1, 2, 3, 4]);
+
+// Test for mixed generic arrays and tuples
+type MixedGenericAndTuple = Array<[number, string]>;
+expectAssignable<ArrayFlat<MixedGenericAndTuple>>([1, 'a', 2, 'b']);
+
+// Test for array with optional elements in nested structure
```
Comment (som-sm) [inline test-d/array-flat.ts:71]: Fails in this case:

```ts
const input: [number, [string, boolean?], number] = [1, ['a'], 2];
const output: ArrayFlat<[number, [string, boolean?], number]> = [1, 'a', 2]; // Errors, but it shouldn't
```

File: type-fest@4.37.0
Comment (som-sm) [inline type-fest@4.37.0:1]: What's this file?

**Review by som-sm — CHANGES_REQUESTED** [review]
@Emiyaaaaa Added comments on the tests — haven’t reviewed the implementation yet.

File: test-d/array-flat.ts (hunk 4 above)
Comment (Emiyaaaaa) [reply test-d/array-flat.ts:71]: `ArrayFlat<[number, [string, boolean?], number]>` should be `[number, string, boolean?, number?]` or `[number, string, number] | [number, string, boolean, number]`

Comment (som-sm) [reply test-d/array-flat.ts:71]: @Emiyaaaaa 
> `[number, string, boolean?, number?]`

This isn't correct, consider this example:
```ts
const input: [number, [string, boolean?], number] = [1, ['a'], 2];
const output: [number, string, boolean?, number?] = [1, 'a', 2]; // Errors, but it shouldn't
```

<br>

> `[number, string, number] | [number, string, boolean | undefined, number]` 

This feels correct to me. Would be nice if the `| undefined` is present only when `exactOptionalPropertyTypes` is disabled.

Comment (Emiyaaaaa) [reply test-d/array-flat.ts:71]: > This feels correct to me
I agree

> Would be nice if the | undefined is present only when exactOptionalPropertyTypes is disabled.

I think `| undefined` is necessary.
```ts
const input: [number, [string, boolean?], number] = [1, ['a', undefined], 2];
const output: [number, string, boolean, number] = [1, 'a', undefined, 2]; // Errors, but it shouldn't

```

Comment (som-sm) [reply test-d/array-flat.ts:71]: > I think | undefined is necessary.

But if `exactOptionalPropertyTypes` is enabled, the `input` itself would error. `undefined` for optional indices will _only_ be allowed if `exactOptionalPropertyTypes` is disabled.

```ts
// @exactOptionalPropertyTypes: true

const input: [number, [string, boolean?], number] = [1, ['a', undefined], 2]; // Errors
```

Comment (Emiyaaaaa) [reply test-d/array-flat.ts:71]: Okay I got, but there seems to be no way to know `exactOptionalPropertyTypes` value

Comment (som-sm) [reply test-d/array-flat.ts:71]: Yeah, there isn't a way to know the value of `exactOptionalPropertyTypes`.

But in some cases the compiler automatically decides whether to add `| undefined` or not depending on the value of `exactOptionalPropertyTypes`, like:

```ts
// @exactOptionalPropertyTypes: false

type Input = [number, boolean?]

type Nullify<T> = {
    [P in keyof T]: T[P] | null
}

type T = Nullify<Input>
//   ^? type T = [number | null, (boolean | null | undefined)?]
```

```ts
// @exactOptionalPropertyTypes: true

type Input = [number, boolean?]

type Nullify<T> = {
    [P in keyof T]: T[P] | null
}

type T = Nullify<Input>
//   ^? type T = [number | null, (boolean | null)?]
```

But don't worry too much about this—just see once if the compiler is automatically doing the trick, if not, it's okay to explicitly add the `| undefined`.

Comment (Emiyaaaaa) [reply test-d/array-flat.ts:71]: Thanks you for your inspiration, I find a way to get `exactOptionalPropertyTypes` option value:
```ts
type ExactOptionalPropertyTypesEnable = [(string | undefined)?] extends [string?] ? false : true;
```

Comment (som-sm) [reply test-d/array-flat.ts:71]: @Emiyaaaaa That's pretty cool ✨

**Comment by sindresorhus** [comment]
@Emiyaaaaa Can you fix the conflicts?

**Comment by Emiyaaaaa** [comment]
> @Emiyaaaaa Can you fix the conflicts?

done

File: source/array-flat.d.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+Return true if the number is 0
+*/
+type IsZero<T extends number> = [T] extends [0] ? true : false;
+
+/**
+Options for the `ArrayFlat` type.
+*/
+type ArrayFlatOptions = {
+	/**
+	 * The number of times to repeat the array items when flattening an non-fixed length array.
```
Comment (sindresorhus) [inline source/array-flat.d.ts:20]: Don't prefix the doc comments with `*`

File: source/array-flat.d.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	 * ```
+	 * type FlatArr0 = ArrayFlat<Array<number, string>, 1, { maxRepeat: 3 }>;
+	 * //=> type FlatArr0 =
+	 * []
+	 * | [number, string]
+	 * | [number, string, number, string]
+	 * | [number, string, number, string, number, string];
+	 * ```
+	 *
+	 * @default 5
```
Comment (sindresorhus) [inline source/array-flat.d.ts:32]: Default should be before example

File: source/array-flat.d.ts [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+};
+
+type DefaultArrayFlatOptions = {
+	maxRepeat: 5;
+};
+
+/**
+Creates a new array type by flattening an array to a specified depth.
+
+Use-case: Flatten an array type to a specified depth.
```
Comment (sindresorhus) [inline source/array-flat.d.ts:44]: This just repeats the description. Provide a real-world use-case.

**Comment by sindresorhus** [comment]
I recommend copy-pasting the type into ChatGPT and ask it to "review, find bugs, and suggest improvements".

File: source/internal/array.d.ts
Comment (som-sm) [inline source/internal/array.d.ts:1]: Also add tests for these helpers. 

And, if these helpers are very specific to `ArrayFlat` then it doesn't make sense to put them here. Because, for example, if I see something like `IsLeadingSpreadArray` as a helper, I'd expect it to work with all types of arrays, not just a specific subset.

**Review by som-sm — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
@Emiyaaaaa Can you fix the merge conflict? (Don't overwrite my commits)

File: test-d/array-flat.ts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+// Test with readonly arrays
+expectType<ArrayFlat<readonly [1, readonly [2, 3]]>>([1, 2, 3]);
+expectType<ArrayFlat<readonly [readonly [1, 2], readonly [3, 4]]>>([1, 2, 3, 4]);
+
+// Edge cases
+expectType<ArrayFlat<Array<[]>>>([]);
+expectType<ArrayFlat<[[]]>>([]);
+expectType<ArrayFlat<[[], []]>>([]);
+expectType<ArrayFlat<undefined[]>>(null! as undefined[]);
+expectType<ArrayFlat<unknown[]>>(null! as unknown[]);
```
Comment (sindresorhus) [inline test-d/array-flat.ts:101]: some more:

```ts
expectType<ArrayFlat<any>>(null! as any);
expectType<ArrayFlat<never>>(null! as never);
expectType<ArrayFlat<any[]>>(null! as any[]);
expectType<ArrayFlat<never[]>>(null! as never[]);
expectType<ArrayFlat<[any, [never]]>>(null! as [any, never]);
```

**Comment by sindresorhus** [comment]
Some things:

- Depth handling does not currently match Array#flat for negative and non-integer depths. In JS, depth is normalized, so values like -1 and 0.5 should effectively behave like 0 (no flattening). Right now those cases can still flatten, and fractional depth can also trigger excessive type instantiation.

- A few tuple type tests have extra internal whitespace before ].

File: test-d/array-flat.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+type MixedDepthArray = [number, string[], [[boolean]]];
+expectAssignable<ArrayFlat<MixedDepthArray>>([1, 'a', 'b', [true]]);
+
+// Test for readonly nested arrays with different depths
+type ReadonlyNestedComplex = readonly [number, ReadonlyArray<readonly string[]>];
+expectAssignable<ArrayFlat<ReadonlyNestedComplex, 2>>([1, 'a', 'b', 'c']);
+
+// Test for recursive flattening with non-array elements
+type RecursiveWithNonArray = [number, [string, {a: number}]];
+expectAssignable<ArrayFlat<RecursiveWithNonArray>>([1, 'string', {a: 42}]);
```
Comment (sindresorhus) [inline test-d/array-flat.ts:157]:
```suggestion
expectAssignable<ArrayFlat<RecursiveWithNonArray>>([1, 'string', {a: 42}]);

// Test options.maxRepeat for non-fixed length arrays
type RepeatedPair = ArrayFlat<Array<[number, string]>, 1, {maxRepeat: 2}>;
expectAssignable<RepeatedPair>([]);
expectAssignable<RepeatedPair>([1, 'a']);
expectAssignable<RepeatedPair>([1, 'a', 2, 'b']);
expectNotAssignable<RepeatedPair>([1]);
expectNotAssignable<RepeatedPair>([1, 'a', 2]);
expectNotAssignable<RepeatedPair>([1, 'a', 2, 'b', 3, 'c']);

type SingleRepeatedPair = ArrayFlat<Array<[number, string]>, 1, {maxRepeat: 1}>;
expectAssignable<SingleRepeatedPair>([]);
expectAssignable<SingleRepeatedPair>([1, 'a']);
expectNotAssignable<SingleRepeatedPair>([1, 'a', 2, 'b']);

type NoRepeatedPair = ArrayFlat<Array<[number, string]>, 1, {maxRepeat: 0}>;
expectType<NoRepeatedPair>([]);
expectNotAssignable<NoRepeatedPair>([1, 'a']);
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Doesn't seem to handle this: https://github.com/sindresorhus/type-fest/pull/1085/commits/302efd8962e0756c079599ed62156ed5d53ce66f Which the runtime version handles.

**Comment by Emiyaaaaa** [comment]
> Doesn't seem to handle this: [302efd8](https://github.com/sindresorhus/type-fest/commit/302efd8962e0756c079599ed62156ed5d53ce66f) Which the runtime version handles.

I could implement like `math.floor(depth)` and `math.max(0, depth)` on depth, but I feel it's over design, almost no one would attempt to use decimals or negative values in the depth parameter.

**Comment by sindresorhus** [comment]
You are probably right. I think the tests are still valuable to show behavior though, just fix them up to pass.

**Comment by Emiyaaaaa** [comment]
> You are probably right. I think the tests are still valuable to show behavior though, just fix them up to pass.

and another reason is this would add extra performance overhead. I don’t think it’s worth introducing guaranteed performance issues just for something that almost never happens.
